File: javascript/json-transform/src/functions/common/ObjectFunctionContext.ts
import FunctionContext from "./FunctionContext";
import { isMap, isNullOrUndefined, toObjectFieldPath } from "../../JsonHelpers";

class ObjectFunctionContext extends FunctionContext {
  private definition: any;
  private constructor(path: string, definition: any, functionKey: string, func: any, resolver: any, extractor: any) {
    super(path, functionKey, func, resolver, extractor);
    this.definition = definition;
  }

  public static async createAsync(
    path: string,
    definition: any,
    functionKey: string,
    func: any,
    resolver: any,
    extractor: any,
  ) {
    let objResolver = resolver;
    if (definition?.[FunctionContext.CONTEXT_KEY]) {
      const contextElement = definition[FunctionContext.CONTEXT_KEY];
      if (isMap(contextElement)) {
        objResolver = await FunctionContext.recalcResolver(path, contextElement, resolver, extractor);
      }
    }
    return new ObjectFunctionContext(path, definition, functionKey, func, objResolver, extractor);
  }

  override has(name: string): boolean {
    return Object.prototype.hasOwnProperty.call(this.definition, name);
  }

  override async get(name: string, transform: boolean = true): Promise<any> {
    const el = this.definition[name == null ? this.alias : name];
    if (isNullOrUndefined(el)) {
      return this.function.getDefaultValue(name);
    }
    return !transform ? el : await this.extractor.transform(this.getPathFor(name), el, this.resolver, true);
  }

  override getPathFor(key: number | string | null): string {
    return this.path + (typeof key === "number" ? `[${key}]` : toObjectFieldPath(!key ? this.getAlias() : key));
  }
}

export default ObjectFunctionContext;

File: java/json-transform/src/main/java/co/nlighten/jsontransform/functions/common/ObjectFunctionContext.java
package co.nlighten.jsontransform.functions.common;

import co.nlighten.jsontransform.JsonTransformerUtils;
import co.nlighten.jsontransform.JsonTransformerFunction;
import co.nlighten.jsontransform.ParameterResolver;
import co.nlighten.jsontransform.adapters.JsonAdapter;

public class ObjectFunctionContext extends FunctionContext {
    private final Object definition;

    public ObjectFunctionContext(String path,
                                 Object definition,
                                 JsonAdapter<?, ?, ?> adapter,
                                 String functionKey,
                                 TransformerFunction function,
                                 ParameterResolver resolver, JsonTransformerFunction extractor) {
        super(path, adapter, functionKey, function, resolver, extractor, definition);
        this.definition = definition;
    }

    @Override
    public boolean has(String name) {
        return adapter.has(definition, name);
    }

    @Override
    public Object get(String name, boolean transform) {
        var el = adapter.get(definition, name == null ? alias : name);
        if (adapter.isNull(el)) {
            return function.getDefaultValue(name);
        }
        return transform ? extractor.transform(getPathFor(name), el, resolver, true) : el;
    }

    @Override
    public String getPathFor(String name) {
        return path + JsonTransformerUtils.toObjectFieldPath(adapter, name == null ? getAlias() : name);
    }
}
